lib/freebl/algcmac.c

```diff
@@ -0,0 +1,324 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#ifdef FREEBL_NO_DEPEND
+#include "stubs.h"
+#endif
+
+#include "secport.h"
+#include "blapit.h"
+#include "blapi.h"
+#include "algcmac.h"
+#include "secerr.h"
+#include "nspr.h"
+
+#include "stdio.h"
+
+// Enum for identifying the underlying block cipher we're using internally.
+// To add a new cipher, add an entry here, define new CMAC_{name}Init and
+// CMAC_{name}Create methods, add a new Context pointer to the cipher union,
+// and update the cmac_Encrypt and CMAC_Destroy methods to handle the new
+// cipher.
+typedef enum {
+    CMAC_AES = 0
+} CMACCiphers;
+
+struct CMACContextStr {
+    // Information about the block cipher to use internally. The cipher should
+    // be placed in ECB mode so that we can use it raw. Store a pointer to the
+    // cipher's context, and keep track of the block size internally.
+    CMACCiphers cipherType;
+    union {
+        void *raw;
+        AESContext *aes;
+    } cipher;
+    int blockSize;
+
+    // Whether or not we allocated (via CMAC_Create) this CMACContext.
+    PRBool wasAllocated;
+
+    // Internal keys which are conditionally used by the algorithm. Derived
```

Comment:
Not keeping the original key was a design decision: let AES deal with all of its keys. This prevents cloning as AES does not expose a context Clone mechanism.